internal/infrastructure/kubernetes/ratelimit/resource_provider.go

```diff
@@ -294,12 +297,105 @@ func (r *ResourceRender) Deployment() (*appsv1.Deployment, error) {
+	selector := resource.GetSelector(rateLimitLabels())
+
+	podLabels := rateLimitLabels()
+	if r.rateLimitDaemonset.Pod.Labels != nil {
+		maps.Copy(podLabels, r.rateLimitDaemonset.Pod.Labels)
+		// Copy overwrites values in the dest map if they exist in the src map https://pkg.go.dev/maps#Copy
+		// It's applied again with the rateLimitLabels that are used as deployment selector to ensure those are not overwritten by user input
+		maps.Copy(podLabels, rateLimitLabels())
+	}
+
+	var podAnnotations map[string]string
+	if enablePrometheus(r.rateLimit) {
+		podAnnotations = map[string]string{
+			"prometheus.io/path":   "/metrics",
+			"prometheus.io/port":   strconv.Itoa(PrometheusPort),
+			"prometheus.io/scrape": "true",
+		}
+	}
+	if r.rateLimitDaemonset.Pod.Annotations != nil {
+		if podAnnotations != nil {
+			maps.Copy(podAnnotations, r.rateLimitDaemonset.Pod.Annotations)
+		} else {
+			podAnnotations = r.rateLimitDaemonset.Pod.Annotations
+		}
+	}
+
+	daemonset := &appsv1.DaemonSet{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       ResourceKindDaemonset,
+			APIVersion: appsAPIVersion,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: r.Namespace,
+			Labels:    rateLimitLabels(),
+		},
+		Spec: appsv1.DaemonSetSpec{
+			UpdateStrategy: *r.rateLimitDaemonset.Strategy,
+			Selector:       selector,
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:      podLabels,
+					Annotations: podAnnotations,
+				},
+				Spec: corev1.PodSpec{
+					Containers:                    containers,
+					ServiceAccountName:            InfraName,
+					AutomountServiceAccountToken:  ptr.To(false),
+					TerminationGracePeriodSeconds: ptr.To[int64](300),
+					DNSPolicy:                     corev1.DNSClusterFirst,
+					RestartPolicy:                 corev1.RestartPolicyAlways,
+					SchedulerName:                 "default-scheduler",
+					SecurityContext:               r.rateLimitDaemonset.Pod.SecurityContext,
+					Volumes:                       expectedDeploymentVolumes(r.rateLimit, r.rateLimitDaemonset.Pod),
+					Affinity:                      r.rateLimitDaemonset.Pod.Affinity,
+					Tolerations:                   r.rateLimitDaemonset.Pod.Tolerations,
+					ImagePullSecrets:              r.rateLimitDaemonset.Pod.ImagePullSecrets,
+					NodeSelector:                  r.rateLimitDaemonset.Pod.NodeSelector,
+				},
+			},
+		},
+	}
+
+	// set name
+	if r.rateLimitDaemonset.Name != nil {
+		daemonset.ObjectMeta.Name = *r.rateLimitDaemonset.Name
+	} else {
+		daemonset.ObjectMeta.Name = r.Name()
+	}
+
+	if r.ownerReferenceUID != nil {
+		if uid, ok := r.ownerReferenceUID[ResourceKindDaemonset]; ok {
+			daemonset.OwnerReferences = []metav1.OwnerReference{
+				{
+					Kind:       ResourceKindDaemonset,
+					APIVersion: appsAPIVersion,
+					Name:       "envoy-gateway",
+					UID:        uid,
+				},
+			}
+		}
+	}
+
+	// apply merge patch to deployment
```

Comment:
apply merge patch to daemonset

Reply:
I'll do another check on the rest of the comments referencing `deployment` now

Reply:
Fixed this one and cleaned up some otherse